source/objects/RenderTarget.cpp

```diff
@@ -209,6 +216,197 @@ Pixel RenderTarget::SampleBilinear(float u, float v, uint8_t mip) const
+			if (imageData == nullptr) return false;
+
+			uint8_t mips = 1;
+			if (generateMips) {
+				mips = MipsFromDimensions(imageWidth, imageHeight);
+			}
+
+			// Resize the RT to fit the image.
+			bool resized = Resize(imageWidth, imageHeight, mips);
+			if (!resized) {
+				stbi_image_free(imageData);
+				return false;
+			}
+
+			memcpy(reinterpret_cast<void*>(GetRawData()), reinterpret_cast<void*>(imageData), mPixelSize * imageWidth * imageHeight);
+			stbi_image_free(imageData);
+		}
+		break;
+	case RTFormat::RGB888: // Unsigned char RTs
+	case RTFormat::RG88:
+	case RTFormat::R8:
+		{
+			stbi_uc* imageData = stbi_load(newPath, &imageWidth, &imageHeight, nullptr, channels);
+			// Resize the RT to fit the image.
+
+			uint8_t mips = 1;
+			if (generateMips) {
+				mips = MipsFromDimensions(imageWidth, imageHeight);
+			}
+
+			bool resized = Resize(imageWidth, imageHeight, mips);
+			if (!resized) {
+				stbi_image_free(imageData);
+				return false;
+			}
+
+			memcpy(reinterpret_cast<void*>(GetRawData()), reinterpret_cast<void*>(imageData), mPixelSize * imageWidth * imageHeight);
+			stbi_image_free(imageData);
+		}
+		break;
+	default:
+		return false;
+		break;
+	}
+
+	return true;
+}
+
+bool RenderTarget::Save(const char* filename, uint8_t mip)
+{
+	int channels = CHANNELS[static_cast<uint8_t>(mFormat)];
+	if (mip >= GetMIPs()) {
+		return false;
+	}
+
+	fs::path filepath(filename);
+	fs::path workingDir = fs::current_path();
+
+	workingDir += "/garrysmod/data";
+	workingDir = fs::absolute(workingDir);
+
+	const bool isFloat = mFormat == RTFormat::RGBFFF || mFormat == RTFormat::RGFF || mFormat == RTFormat::RF;
+
+	std::string extension = isFloat ? "hdr" : "png";
+
+	if (filepath.has_extension()) {
+		std::string ext = filepath.extension().string();
+		ext = ext.substr(1); // Extension returns the period along with the actual extension.
+
+		if (ext == "hdr" || ext == "png" || ext == "jpg" || ext == "bmp") {
+			extension = ext; // Change the default to the requested one
+		}
+		else {
+			// Append the default extension.
+			// Great for things like: "frame.001"
+			filepath += "." + extension;
+		}
+	}
+
+	fs::path finalPath = workingDir / filepath;
+	finalPath = fs::absolute(finalPath);
+	finalPath = finalPath.make_preferred();
+
+	if (!verifyInRoot(finalPath, workingDir)) {
+		return false;
+	}
+	
+	if (filepath.has_parent_path()) {
+		// Create the subdirectories the user wants.
+		fs::create_directories(finalPath.parent_path());
+	}
+
+	const char* rawFilepath = finalPath.string().c_str();
+	
+	uint16_t width = GetWidth();
+	uint16_t height = GetHeight();
+
+	if (mip > 0) {
+		width >>= mip;
+		height >>= mip;
+
+		// Check for less than 0 (happens on non-base2 resolutions)
+		width = width < 1 ? 1 : width;
+		height = height < 1 ? 1 : height;
+	}
+
+	if (extension == "hdr") {
+		if (!stbi_write_hdr(rawFilepath, width, height, channels, reinterpret_cast<const float*>(GetRawData(mip)))) {
```

Comment:
You're assuming that the contents of this RT are floats, despite letting the user specify the `hdr` file format on other RT formats